### ISSUE TITLE
Scaling - Running multiple NBXplorer instances in a cluster

### ISSUE THREAD

**Issue by sigHash** [dgarage/NBXplorer#82]
Just wondering what are the best practices for running multiple instances? Can NBXplorer scale horizontally?
Thanks

**Comment by NicolasDorier** [comment]
No, NBXplorer is just saving stuff inside an embedded NoSql database. (dbreeze)

That said, things would work if you rewrite the Repository class to use a centralized clustered database. (Doing that for another project)

**Comment by sigHash** [comment]
Makes sense. Any plan for making that change open source?

**Comment by NicolasDorier** [comment]
I don't think my company would want it.
The interface of Repository is quite simple actually. Take a look at the public methods. 

Extract them inside an interface. Then implement this interface.
The interface is quite easy. The current DBreeze implementation is quite messy because of back compatiblity stuff.

**Comment by sigHash** [comment]
Any recommendation for the type of DB to use? Was thinking of Postgres

**Comment by NicolasDorier** [comment]
I love PostgreSQL. I advise you to not use Entity Framework, as it add some overhead.

**Comment by sigHash** [comment]
Raw SQL queries then?

**Comment by NicolasDorier** [comment]
Yes, also use 

``CREATE DATABASE blahblah TEMPLATE 'template0' LC_TYPE 'C' ENCODING 'UTF8'`

To create the database. This is necessary so requests on String primary keys does not result in table scan.

**Comment by NicolasDorier** [comment]
Another idea you can have, is keeping `Repository` as is, but abstracting the something inside to keep the already written logic.

**Comment by sigHash** [comment]
Didn't quite get the second thought

**Comment by NicolasDorier** [comment]
Ah just forget, extracting an interface out of IRepository is easier.

**Comment by sigHash** [comment]
So forgetting the already written public methods?

**Comment by NicolasDorier** [comment]
Check public methods of Repository, extract them as interface, and do your own implementation using Postgresql. I can cleanup things later.

**Comment by sigHash** [comment]
Sounds like a plan

**Comment by NicolasDorier** [comment]
Can start tomorrow? I plan doing some refactoring which might make your life easier.

**Comment by sigHash** [comment]
I can start in few days. Would love to have those changes

**Comment by farukterzioglu** [comment]
I would like to revisit this issue. 
I am thinking about possibilities of using a relational db (like ms sql) for nbx.

After 2 years, what do you think @NicolasDorier about this.

I just check repository class and saw that some business logic (like needRemove list or needUpdate list) and db related codes (DBriize) are side by side. 

It may be hard to abstract db related codes. 

But I am thinking about this seriously since getting utxo list sometimes takes more than 2 minutes. 
(also having another nbx instance as a backup is a real pain)
(This is another topic to discuss.)

I would like to know  what are the things I need to take a look if I wish to write another repository to work with for example ms sql

**Comment by NicolasDorier** [comment]
@farukterzioglu in a proprietary fork of NBXplorer, I did it. You can do by abstracting the "Index" classes in the repository.

> getting utxo list sometimes takes more than 2 minutes.

If this is the problem you are having, using a DB instead won't solve the issue. (even if on another server, as latency might be problematic)
